BUG: Fix copy semantics in ``__array__``

seberg opened the pull request:
This should fix the semantics of ``__array__``.  While rejecting ``copy=False`` is OK even if unnecessary, ``copy=True`` should never have been ignored and is dangerous.

Closes gh-57739, closes gh-59932, closes https://github.com/pandas-dev/pandas/issues/59614

- [x] Still needs new tests.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

----

I have to figure out the tests.  I think a test in form of:
```
arr1 = np.asarray(obj, copy=True)
arr2 = np.asarray(obj, copy=True)
assert not np.may_share_memory(arr1, arr2)
# Check that without copy always works:
assert_array_equal(arr1, np.asarray(obj))

if np_ver < 2:
    return  # copy=False semantics not supported

try:
   arr1 = np.asarray(obj, copy=False)
except ValueError:
   return  # An error is acceptable for `copy=False`

# If no error is given, multiple returns must be views:
arr2 = np.asarray(obj, copy=False)
assert np.may_share_memory(arr1, arr2)
```
will work nicely.  But, right now I am not sure if there is a convenient pattern/parametrization to steal to cover all of the `__array__` implementations here.

seberg committed:
BUG: Fix copy semantics in ``__array__``

This fixes the semantics of ``__array__``.  While rejecting ``copy=False``
is pretty harmless, ``copy=True`` should never have been ignored and is
dangerous.

seberg committed:
BUG: Fix one more path not translating ``copy=`` correctly

seberg committed:
BUG: Avoid asarray with copy= (it was added in 2.0)

seberg committed:
More fixes found by typing checks (or working around them)

seberg commented on pandas/core/arrays/categorical.py:
```diff
@@ -1686,13 +1687,20 @@ def __array__(
+
         ret = take_nd(self.categories._values, self._codes)
-        if dtype and np.dtype(dtype) != self.categories.dtype:
-            return np.asarray(ret, dtype)
```
I did not understand why this is needed.  If dtypes match, NumPy should make it a no-op?  If dtype is None, it is the same as not passing.

jorisvandenbossche replied:
Yeah, also not sure why this is here

seberg commented:
FWIW, if anyone knows how to add decent tests for this, pushing to the PR is appreciated.  Otherwise, I may try to figure out where to add at least tests for a few of these paths.

chaoyihu commented:
Hi, Glad to see a PR addressing #59932 ! I was trying to reproduce that issue but couldn't install NumPy >= 2.0 with the main branch, which seems to be because the `environment.yml` on the current main branch requires  `numpy<2`. How did you resolve it?

And by the way, I asked about the bug fix release on Slack:

> Could this be released as a bug fix for v2.x?

Quoting the reply from @rhshadrach:

> I think there won't be another 2.2.x release; the next release will be 2.3.0.

seberg committed:
TST: Add test for __array__ copy behavior

seberg commented:
Found a test that seemed reasonably to expand.  I doubt it covers _all_ the branches changed, though (and no, I don't love expanding tests, but I was lazy to split out the parametrization).

seberg committed:
TST: Fixup test to use array rather than asarray

asarray did not support `copy=` on older versions of NumPy

seberg commented:
Ping @mroeschke just because it's been a while that it was idle without any tests.  Not sure if this needs a milestoned or a  release note.

mroeschke commented on pandas/core/arrays/categorical.py:
```diff
@@ -1686,13 +1687,20 @@ def __array__(
         >>> np.asarray(cat)
         array(['a', 'b'], dtype=object)
         """
+        if copy is False:
```
If `lib.is_range_indexer(self._codes)` i.e. the `self.categories._values` are all unique then, a copy could be avoided?

seberg replied:
Probably, but I left it out, it doesn't seem vital to try and improve it.  Also fixed things up, because `take_nd` should presumably always return a copy.

mroeschke commented:
Sorry for the delay. Overall looks good! Could you add a note to the `v2.3.0.rst` release notes?

jorisvandenbossche commented on pandas/core/arrays/arrow/array.py:
```diff
@@ -667,7 +667,15 @@ def __array__(
+                "Unable to avoid copy while creating an array as requested."
+            )
+        elif copy is None:
+            copy = False  # The NumPy copy=False meaning is different here.
```
What does this comment exactly mean?

jorisvandenbossche commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -359,7 +359,14 @@ def __array__(
+
+        if copy is True:
+            return np.array(self._ndarray, dtype=dtype)
         return self._ndarray
```
Does it matter much for the numpy side whether we return the ndarray here directly (which might not match the dtype), or already do `return np.asarray(self._ndarray, dtype=dtype)` on our side?

seberg replied:
No, NumPy will deal with dtype mismatch.  It's been historically more of a hint and not a requirement.

jorisvandenbossche commented on pandas/core/arrays/period.py:
```diff
@@ -389,6 +389,11 @@ def freqstr(self) -> str:
     def __array__(
         self, dtype: NpDtype | None = None, copy: bool | None = None
     ) -> np.ndarray:
+        if copy is False:
+            raise ValueError(
+                "Unable to avoid copy while creating an array as requested."
+            )
```
Strictly speaking, the 

```
        if dtype == "i8":
            return self.asi8
```

just below is zero copy, so a `copy=False` is possible in that case

jorisvandenbossche replied:
Ah, and more importantly, that means that that branch is not honoring `copy=True`

seberg replied:
Will look into this and other fixups tomorrow.  Right, wrong `copy=True` is a problem, I definitely err'd on the error side for `copy=False`, so let me change it here (as well).

jorisvandenbossche commented on pandas/core/arrays/sparse/array.py:
```diff
@@ -547,6 +547,11 @@ def from_spmatrix(cls, data: spmatrix) -> Self:
+
         fill_value = self.fill_value
 
         if self.sp_index.ngaps == 0:
```
corner case, but also this branch is returning underlying values and so is not honoring a `copy=True`

jorisvandenbossche commented on pandas/core/generic.py:
```diff
@@ -2015,8 +2015,17 @@ def __array__(
+            arr = np.array(values, dtype=dtype, copy=copy)
+
+        if (
+            copy is not False
```
and also not `True`, because then it will already have copied in `arr = np.array(values, dtype=dtype, copy=copy)` above?

seberg replied:
We pass on `copy=copy`, so we can trust the semantics being correct until here.

seberg replied:
Hmmmm, but I think I got it wrong...  It should rather say, `copy is not True`?  Because if we force a copy it doesn't seem sensible to create a read-only view.

(Have to see if I can write a test for these corrections, it is clearly trickier then just err'ing on the "raise an error" side!)

jorisvandenbossche replied:
Yes, that is what I meant

jorisvandenbossche commented on pandas/core/indexes/multi.py:
```diff
@@ -1391,6 +1391,9 @@ def copy(  # type: ignore[override]
 
     def __array__(self, dtype=None, copy=None) -> np.ndarray:
         """the array interface, return my values"""
+        if copy is True:
+            # Note: branch avoids `copy=None` for NumPy 1.x support
+            return np.array(self.values, dtype=dtype, copy=copy)
```
```suggestion
```

This is unnecessary here, because `MultiIndex.values` is already a customly build array (see `def _values(self)` a bit higher up in this file)

seberg replied:
Ouch, right, in that case we need to raise for ``copy=False`!

jorisvandenbossche reviewed (commented):
Thanks a lot @seberg for working in this! 

Added some more comments (some corner cases (like period converted to int being zero-copy) might need additional specific tests, but also happy to work on those)

seberg committed:
BUG: Fixup ``__array__`` copy paths based on review

A few of these were just wrong, a few others are enhancements to
allow the cases that clearly should work without copy to still pass.

Co-authored-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

seberg commented:
Pushed fixes based on the review, thanks; it's quite tricky to know whether or not a copy may have been made.

Which also means that expandign the tests would be great (with the new changes also with `dtype=` which I am not sure how well it works in the current test).

@jorisvandenbossche would be great if you can look into tests (I can review), since I have to hunt/think a bit more about how to build examples.  But if I don't see movement in the next days, I'll hopefully dive in.

jorisvandenbossche committed:
fix period case and add specific test

jorisvandenbossche committed:
update test to be explicit about copy vs nocopy + allow copy=False for masked arrays in case of no NAs

jorisvandenbossche committed:
add similar test to base extension tests

jorisvandenbossche committed:
add specific test for sparse corner case

jorisvandenbossche committed:
add specific test for MultiIndex

jorisvandenbossche commented:
@seberg thanks for the update! I added a bunch of additional tests, that I think should now cover the corner cases for which you updated the code (i.e. PeriodArray and SparseArray corner cases that allow copy=False, MultiIndex never allowing copy=False). 
Also expanded the existing test a bit and added a basic base extension array test.

jorisvandenbossche commented on pandas/core/arrays/masked.py:
```diff
@@ -507,7 +507,7 @@ def to_numpy(
             with warnings.catch_warnings():
                 warnings.filterwarnings("ignore", category=RuntimeWarning)
-                data = self._data.astype(dtype, copy=copy)
+                data = np.array(self._data, dtype=dtype, copy=copy)
```
Made this change together with allowing to get here with `copy=False`, to ensure if `copy=False` is not actually possible depending on the passed `dtype`, it still properly errors

seberg replied:
This change (together with the `self._hasna` change below) make sense to me.  But unfortunately, `to_numpy(copy=False)` change the meaning of `copy=False` this way.  And I think this is public API?

Maybe better to special case `_hasna` in the branch below?  (could even just ignore `dtype` with a comment that NumPy is fine with that).

jorisvandenbossche replied:
Yeah, will have to update this to not change `to_numpy(copy=False)`, as that is indeed public API we want to keep working that way.

jorisvandenbossche commented on pandas/tests/indexes/multi/test_conversion.py:
```diff
@@ -16,6 +18,40 @@ def test_to_numpy(idx):
+    ]
+    tm.assert_numpy_array_equal(result, expected)
+
+    # it always gives a copy by default, but the values are cached, so results
+    # are still sharing memory
+    result_copy1 = np.asarray(idx)
+    result_copy2 = np.asarray(idx)
+    assert np.may_share_memory(result_copy1, result_copy2)
```
This might be an unfortunate side effect of our caching, because a user might expect here always to get a new object. But at least when explicitly passing `copy=True`, then it appears to do the right thing. But is this numpy that does that? (still copy even if `copy=True` was passed down to `__array__`?) We should probably not keep relying on that, and do an explicit copy on our side in case of `copy=True`?

jorisvandenbossche replied:
It seems I am testing locally with numpy 2.0.2 (so not 1.x, but also not latest 2.1.x or main)

seberg replied:
Ah, maybe there were fixes with 2.1.

seberg commented on pandas/tests/base/test_conversion.py:
```diff
@@ -374,15 +380,16 @@ def test_to_numpy(arr, expected, index_or_series_or_array, request):
+    if not zero_copy:
+        with pytest.raises(ValueError, match="Unable to avoid copy while creating"):
+            # An error is always acceptable for `copy=False`
+            np.array(thing, copy=False)
```
Very nice to make the test specific!

seberg commented on pandas/tests/extension/base/interface.py:
```diff
@@ -71,6 +73,25 @@ def test_array_interface(self, data):
+            result_nocopy1 = np.array(data, copy=False)
+        except ValueError:
+            # An error is always acceptable for `copy=False`
+            return
```
You can make this test specific presumably.

jorisvandenbossche replied:
For this specific case it is harder to make it specific, because we don't know the exact array object and its expected semantics (this is eg also used by external packages implementing an ExtensionArray)

seberg reviewed (commented):
Thanks a lot, the test additions look good to me modulo a nitpick.

I am worried about the change of the meaning for `.to_numpy(copy=False)`.  But maybe this is private API?

If this is private API, then it is OK, but:
* The docs should be changed and both branches should behave the same (i.e. raise an error in the first branch if `copy=False`).
* The default should be `copy=None`.

If not, we can move that special case to the `__array__`.

jorisvandenbossche committed:
fix MultiIndex copy=True case for recent numpy

jorisvandenbossche committed:
fix copy=False case for masked array

seberg commented:
Thanks Joris.  With the last two commits, I think this should be good to go in.

jorisvandenbossche committed:
add whatsnew note

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/main' into array-copy

jorisvandenbossche commented:
The whatsnew note can probably be improved, but going to merge this already so I can start on backporting the code to 2.3.x.

Thanks @seberg!

jorisvandenbossche commented:
Manual backport -> https://github.com/pandas-dev/pandas/pull/60189

jorisvandenbossche commented on pandas/core/series.py:
```diff
@@ -879,8 +879,15 @@ def __array__(
+
+        if copy is True:
+            return arr
+        if copy is False or astype_is_view(values.dtype, arr.dtype):
```
One more issue I noticed while doing the backport: for a similar case in generic.py we changed the `copy is False` to `copy is not True`, which I think we should have done here as well.

Will try to come up with a test case that would fail because of this, and do a follow-up tomorrow.

seberg replied:
Thanks, ouch.  This is clearly hard to get right without tests :/.

The release snippet looks fine to me.

jorisvandenbossche replied:
It also only matters for the `copy=None` case, which AFAIU is quite new, and so we didn't cover that when adding the logic here for making the resulting array read-only

seberg replied:
Looking closer, I think this is right?  Note that it says `copy is False or ...`.  If `copy is False` we don't have to check whether it is a view (because otherwise it would have errored)?

jorisvandenbossche replied:
Ah, that's a good point. I assume the `arr = np.array(values, dtype=dtype, copy=copy)` line above will already have errored in case of `copy=False` if a zero-copy conversion was not possible. 
So indeed the `or` is fine: if `copy=False` we know `arr` is always a view, and for `copy=None` (the one remaining case) we have `astype_is_view` to check it.

Maybe it's then in `generic.py` that we could change the `copy is not True and ..` to `copy is False or ..` (just to use a similar pattern in both cases)

jorisvandenbossche replied:
Looking once more at it, we actually already have tests for this, because we test that we get a read-only array using `np.asarray(ser)`, which essentially means passing `copy=None`. 

I expanded the test with `np.array(ser, copy=False)` to also explicitly cover the `copy=False` case (and ensure this branch setting writeable to False is covered for that case) -> https://github.com/pandas-dev/pandas/pull/60191

jorisvandenbossche commented:
@seberg I am seeing some failures in the geopandas tests using pandas dev because of this change (https://github.com/geopandas/geopandas/actions/runs/11688928152/job/32550520328?pr=3455)

Essentially geopandas is using `np.array(pandas_object, copy=False)` in a few places. And in the past this has always worked, just silently returning a copy (I assume we just passed `copy=False` in the idea of avoiding a copy if not needed, it's not that the code relied on it returning a view). 
I think for numpy itself you made this behaviour change of `np.array` (to start raising an error) on 2.0? Did you warn about it in advance?

I am wondering if we should leave the `copy=False`-raising-error part of the changes for pandas 3.0, instead of including that in pandas 2.3.

The `copy=True` part of the change (actually copying when asked) is I think the main bug fix we definitely should include in 2.3, as far as I understand (I think the issues people have been reporting in https://github.com/pandas-dev/pandas/issues/57739 would be resolved by this part)

seberg commented:
No, there was no warning about it, it was a hard change in 2.0...

You'll be the better judge for pandas users here.  It is wrong to not raise, but it does seem worse to backport `copy=False` for sure, since `copy=False` will be exceedingly rare anyway in code.
If this was right from the start geopandas would have had to fix this as part of NumPy 2 transition, but that ship has sailed.

So I guess, you may only want to backport the `copy=True` fix, since it seems important, but not touch `copy=False` even if it is a bug.
If the next release is 3.0, I guess that is good enough to just do the change without deprecation then.

simonjayhawkins commented:
> I am seeing some failures in the geopandas tests using pandas dev because of this change

@jorisvandenbossche now that this is merged and backported should we have a new issue for this and linked from the 2.3 release discussion for greater visibility?

> I am wondering if we should leave the copy=False-raising-error part of the changes for pandas 3.0, instead of including that in pandas 2.3.

Seems reasonable to me that we probably don't want to introduce breaking changes in 2.3

jorisvandenbossche commented:
Yes, opened https://github.com/pandas-dev/pandas/issues/60340 and labeled it for 2.3.0
